File: python/algos/valid_parentheses.py
def is_left(sign: str) -> bool:
    """Checks if the given sign is left parentheses."""
    return sign == "(" or sign == "{" or sign == "["


def valid_parentheses(expression: str) -> bool:
    """Checks if given expression parentheses are correct."""
    mapping = {")": "(", "]": "[", "}": "{"}
    stack = []
    for sign in expression:
        if is_left(sign):
            stack.append(sign)
        else:
            if not stack or stack[-1] != mapping[sign]:
                return False
            else:
                stack.pop()
    return True


if __name__ == "__main__":
    assert valid_parentheses("()")
    assert valid_parentheses("([][]{}")
    assert not valid_parentheses("([)")

File: rust/src/algos/valid_parentheses.rs
use std::collections::{HashMap, LinkedList};

pub fn valid_parentheses(expression: String) -> bool {
    let mut stack = LinkedList::new();
    let mapping = HashMap::from([(')', '('), ('}', '{'), (']', '[')]);

    for sign in expression.chars() {
        if is_left(sign) {
            stack.push_back(sign);
        } else {
            if stack.is_empty() || stack.back().unwrap() != mapping.get(&sign).unwrap() {
                return false;
            } else {
                stack.pop_back();
            }
        }
    }
    stack.is_empty()
}

fn is_left(sign: char) -> bool {
    sign == '(' || sign == '[' || sign == '{'
}